llama-index-core/llama_index/core/workflow/context.py

```diff
@@ -40,6 +83,83 @@ def __init__(self, workflow: "Workflow", stepwise: bool = False) -> None:
 
+        # keep track of all the event classes that are accepted by the workflow
+        self._event_classes: Dict[str, Type[Event]] = {}
+        for step_func in self._workflow._get_steps().values():
+            step_config: Optional[StepConfig] = getattr(
+                step_func, "__step_config", None
+            )
+            if step_config and step_config.accepted_events:
+                for event_type_cls in step_config.accepted_events:
+                    self._event_classes[event_type_cls.__name__] = event_type_cls
+
+            if step_config and step_config.return_types:
+                for return_type_cls in step_config.return_types:
+                    self._event_classes[return_type_cls.__name__] = return_type_cls
+
+    def _serialize_event(self, event: Event) -> Tuple[str, Any]:
+        return (event.__class__.__name__, event.model_dump())
+
+    def _deserialize_event(self, event_name: str, event_data: Any) -> Event:
+        event_cls = self._event_classes[event_name]
+        return event_cls.model_validate(event_data)
+
+    def _serialize_queue(self, queue: asyncio.Queue) -> str:
+        queue_items = list(queue._queue)  # type: ignore
+        queue_objs = [self._serialize_event(obj) for obj in queue_items]
+        return json.dumps(queue_objs)  # type: ignore
+
+    def _deserialize_queue(self, queue_str: str) -> asyncio.Queue:
+        queue_objs = json.loads(queue_str)
+        queue = asyncio.Queue()  # type: ignore
+        for obj in queue_objs:
+            event_obj = self._deserialize_event(obj[0], obj[1])
+            queue.put_nowait(event_obj)
+        return queue
+
+    def _serialize_globals(self) -> Dict[str, Any]:
+        serialized_globals = {}
+        for key, value in self._globals.items():
+            try:
+                serialized_globals[key] = self.serializer.serialize(value)
+            except Exception as e:
+                raise ValueError(f"Failed to serialize value for key {key}: {e}")
+        return serialized_globals
+
+    def _deserialize_globals(
+        self, serialized_globals: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        deserialized_globals = {}
+        for key, value in serialized_globals.items():
+            try:
+                deserialized_globals[key] = self.serializer.deserialize(value)
+            except Exception as e:
+                raise ValueError(f"Failed to deserialize value for key {key}: {e}")
+        return deserialized_globals
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "globals": self._serialize_globals(),
+            "streaming_queue": self._serialize_queue(self._streaming_queue),
+            "queues": {k: self._serialize_queue(v) for k, v in self._queues.items()},
+            "stepwise": self.stepwise,
+        }
+
+    @classmethod
+    def from_dict(
+        cls,
+        workflow: "Workflow",
+        data: Dict[str, Any],
+        serializer: Optional[ContextSerializer] = None,
```

Comment:
since we're here, why not passing a serializer instance to `to_dict` as well, so we avoid passing it to the constructor? This would have the advantage of being able to serialize the same workflow instance with two different serializers without creating a new one.

Reply:
true!

Reply:
done